### ISSUE TITLE
Disabling unnecessary detectors in the Sonar plugin

### ISSUE THREAD

**Issue by gtoison** [spotbugs/spotbugs#1898]
Hello, I'm trying to reduce the Sonar SpotBugs analysis time by disabling unnecessary detectors. 
For instance if the rules `IL_INFINITE_RECURSIVE_LOOP` and `IL_CONTAINER_ADDED_TO_ITSELF` are not enabled in SonarQube, then it should not be necessary to run the `edu.umd.cs.findbugs.detect.InfiniteRecursiveLoop` detector.

I've seen that some detectors are "reporting" detectors and also there's an execution plan because some detectors have constraints to run before/after other detectors. For instance `edu.umd.cs.findbugs.detect.CalledMethods` is constrained to run before `edu.umd.cs.findbugs.detect.UncallableMethodOfAnonymousClass`, so I suppose that `UncallableMethodOfAnonymousClass` needs some results computed by `CalledMethods`?

In that case is there a way to restrict the detectors based on a list of reported bug patterns? 

I looked into the code of the Eclipse plugin and came up with the below code, but I'm not sure if I could be missing something important.

```java
  private void disableUnnecessaryDetectors(UserPreferences userPreferences) {
    for (DetectorFactory detectorFactory : DetectorFactoryCollection.instance().getFactories()) {
      boolean enabled = !detectorFactory.isReportingDetector() || detectorFactoryHasActiveRules(detectorFactory);
      
      userPreferences.enableDetector(detectorFactory, enabled);
    }
  }
  
  private boolean detectorFactoryHasActiveRules(DetectorFactory detectorFactory) {
    Collection<String> repositories = repositoriesForPlugin(detectorFactory.getPlugin());
    
    if (repositories.isEmpty()) {
      LOG.warn("Detector {} is activated because it is not from a built-in plugin, cannot check if there are some active rules", detectorFactory);
      return true;
    }
    
    for (BugPattern bugPattern : detectorFactory.getReportedBugPatterns()) {
      String bugPatternType = bugPattern.getType();
      
      for (String repository : repositories) {
        RuleKey ruleKey = RuleKey.of(repository, bugPatternType);

        if (activeRules.find(ruleKey) != null) {
          return true;
        }
      }
    }
    
    return false;
  }

  /**
   * @param plugin
   * @return a collection of repository keys or an empty collection in case the plugin is unknown
   */
  private Collection<String> repositoriesForPlugin(Plugin plugin) {
    switch (plugin.getPluginId()) {
    case FINDBUGS_CORE_PLUGIN_ID:
      return Collections.singletonList(FindbugsRulesDefinition.REPOSITORY_KEY);
    case FINDSECBUGS_PLUGIN_ID:
      return Arrays.asList(
          FindSecurityBugsRulesDefinition.REPOSITORY_KEY, 
          FindSecurityBugsJspRulesDefinition.REPOSITORY_KEY);
    case SPOTBUGS_CONTRIB_PLUGIN_ID:
      return Collections.singletonList(FbContribRulesDefinition.REPOSITORY_KEY);
    default:
      return Collections.emptyList();
    }
  }
```
